### PR TITLE
[12.0] Add module server_environment_data_encryption

### PR DESCRIPTION
**Pull request by florian-dacosta**
This module aims to replace keychain.
At the start keychain came from the need to manage different value for different environments (like server_environment) but making the configuration possible by the user (directly in the UI and not in configuration file.) The second need was to have encrypted data, for example, to avoid giving critical information hen you give a backup for developement purpose.

The main issue is that not everyone want to use keychain (Server_environement is used a lot in the community) and so it can't be a dependencie in some base module, which forces to do a lot of 'glue' module each time we need it.

Since the recent great improvements made in server_environment, we can now do something compatible.
So this module depends on server_environment and allow to configure data by environements, directly in the User Interface and encrypted in the database.

First, odoo will try to get the values from server_environement, if it gets nothing, it will fallback on the encrypted data in the database.
Also, it allows user to configure the data for all the environement from the place.
For instance, if you install mail_environment and this module, from the incoming mail server form, you have smart buttons to switch to dev environment and set values, then for preproduction environment and set the values, and the production environement, of course.
So it is really easy for a normal user to make these configuration if necessary.

Depends on https://github.com/OCA/server-env/pull/23

@sebastienbeau @hparfr @guewen @lmignon

### CODE REVIEW

**Comment by rvalyi** [comment]
Hello @florian-dacosta 
I would advocate for using LGPL3 here. I think the LGPL part of Odoo is getting large (we can like it or not but that's a fact). But putting such base infrastructure as AGPL, I think the only thing we do is limiting its adoption and making it likely a competing LGPL solution to appear and take over this module no matter how good it is. We would fool ourselves thinking people would suddenly open their codebase instead of rewritting such a module. So personally I think it is still good that the OCA keep fighting for AGPL on more sophisticated ERP modules for which openness is the major selling point but for basic infrastructure modules like this one that could be appealing for startups using Odoo as a framework I think we better think twice. BTW I think the same about base_jsonify and a few other Akretion framework modules.

**Comment by sebastienbeau** [comment]
Hi on my side I always prefers to use AGPLv3 first and then see if somebody ask for changing it.
Moreover server env is not LGPL so AGPL is good

**Review by gurneyalex — APPROVED** [review]

File: server_environment_data_encryption/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+{
+    "name": "Server Environment Data Encryption",
+    "version": "12.0.1.0.0",
+    "development_status": 'Alpha',
+    "category": "Tools",
+    "website": "https://akretion.com/",
```
Comment (simahawk) [inline server_environment_data_encryption/__manifest__.py]: https://github.com/OCA/server-env

File: server_environment_data_encryption/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    "author": "Akretion, Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "application": False,
```
Comment (simahawk) [inline server_environment_data_encryption/__manifest__.py]: no need for application flag

File: server_environment_data_encryption/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    "installable": True,
+    "depends": ["server_environment", "data_encryption"],
+    "data": [],
```
Comment (simahawk) [inline server_environment_data_encryption/__manifest__.py]: no need for data

File: server_environment_data_encryption/models/server_env_mixin.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+
+_logger = logging.getLogger(__name__)
+
+
+class ServerEnvMixin(models.AbstractModel):
+    _inherit = "server.env.mixin"
+
+    def _compute_server_env_from_default(self, field_name, options):
+        "First return database encrypted value then default value"
```
Comment (simahawk) [inline server_environment_data_encryption/models/server_env_mixin.py]: triple quotes

File: server_environment_data_encryption/models/server_env_mixin.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+                values.update(new_val)
+                encrypted_data_obj._encrypted_store_json(
+                    encrypted_data_name, values, env=env
+                )
+
+    def action_change_env_data_encrypted_fields(self):
+        action_id = self.env.context.get("params", {}).get("action")
+        if not action_id:
+            # We don't know which action we are using... take one random.
+            action_id = self.env['ir.actions.act_window'].search(
```
Comment (simahawk) [inline server_environment_data_encryption/models/server_env_mixin.py]: you can use `self.get_formview_action`

File: server_environment_data_encryption/models/server_env_mixin.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        action["res_id"] = self.id
+        views_form = []
+        for view_id, view_type in action.get("views", []):
+            if view_type == "form":
+                views_form.append((view_id, view_type))
+        action["views"] = views_form
+        return action
+
+    def _get_extra_environment_info_div(self, current_env, extra_envs):
+        button_div = "<div>"
```
Comment (simahawk) [inline server_environment_data_encryption/models/server_env_mixin.py:73]: would nicer and easier to use a qweb template to generate this.
Not a blocker but worth a TODO comment.

File: server_environment_data_encryption/tests/test_server_environment_data_encrypt.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo.addons.data_encryption.tests.common import CommonDataEncrypted
+from pathlib import Path
+
+
+class TestServerEnvDataEncrypted(CommonDataEncrypted):
+
+    #    def test_store_data_no_superuser(self):
```
Comment (simahawk) [inline server_environment_data_encryption/tests/test_server_environment_data_encrypt.py]: needed?

**Review by simahawk — CHANGES_REQUESTED** [review]

File: server_environment_data_encryption/__manifest__.py (hunk 1 above)
Comment (florian-dacosta) [reply server_environment_data_encryption/__manifest__.py]: Done

File: server_environment_data_encryption/__manifest__.py (hunk 2 above)
Comment (florian-dacosta) [reply server_environment_data_encryption/__manifest__.py]: Done

File: server_environment_data_encryption/__manifest__.py (hunk 3 above)
Comment (florian-dacosta) [reply server_environment_data_encryption/__manifest__.py]: Done

File: server_environment_data_encryption/models/server_env_mixin.py (hunk 4 above)
Comment (florian-dacosta) [reply server_environment_data_encryption/models/server_env_mixin.py]: Done

File: server_environment_data_encryption/models/server_env_mixin.py (hunk 5 above)
Comment (florian-dacosta) [reply server_environment_data_encryption/models/server_env_mixin.py]: I didn't know this one, thanks for the pointer, I made the change.

File: server_environment_data_encryption/tests/test_server_environment_data_encrypt.py (hunk 7 above)
Comment (florian-dacosta) [reply server_environment_data_encryption/tests/test_server_environment_data_encrypt.py]: no, I removed it.

File: server_environment_data_encryption/models/server_env_mixin.py (hunk 6 above)
Comment (florian-dacosta) [reply server_environment_data_encryption/models/server_env_mixin.py:73]: I put a TODO comment, I can't do this now

**Review by simahawk — APPROVED** [review]
builds are :red_circle:

**Comment by florian-dacosta** [comment]
Yes, the module data_encryption is not merged yet and it is a dependency

**Comment by sebastienbeau** [comment]
I have rebased the PR so test should be green and ready to merge

**Comment by sebastienbeau** [comment]
let's merge it \o/ /ocabot merge

**Pull request merged**